src/transformers/models/focalnet/modeling_focalnet.py

```diff
@@ -0,0 +1,992 @@
+                )
+                for i in range(depth)
+            ]
+        )
+
+        if downsample is not None:
+            self.downsample = downsample(
+                image_size=input_resolution,
+                patch_size=2,
+                num_channels=dim,
+                embed_dim=out_dim,
+                add_norm=True,
+                use_conv_embed=config.use_conv_embed,
+                is_stem=False,
+            )
+        else:
+            self.downsample = None
+
+        self.pointing = False
+
+    def forward(self, hidden_states: torch.Tensor, input_dimensions: Tuple[int, int]) -> Tuple[torch.Tensor]:
+        height, width = input_dimensions
+        for layer_module in self.layers:
+            hidden_states = layer_module(hidden_states, input_dimensions)
+
+        hidden_states_before_downsampling = hidden_states
+        if self.downsample is not None:
+            height, width = input_dimensions
+            hidden_states = hidden_states.transpose(1, 2).reshape(
+                hidden_states_before_downsampling.shape[0], -1, height, width
+            )
+            hidden_states, output_dimensions = self.downsample(hidden_states)
+
+        else:
+            output_dimensions = (height, width, height, width)
+
+        stage_outputs = (hidden_states, hidden_states_before_downsampling, output_dimensions)
+
+        return stage_outputs
+
+
+class FocalNetEncoder(nn.Module):
+    def __init__(self, config, grid_size):
+        super().__init__()
+        self.num_stages = len(config.depths)
+        self.config = config
+
+        # stochastic depth
+        dpr = [
+            x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))
+        ]  # stochastic depth decay rule
+
+        embed_dim = [config.embed_dim * (2**i) for i in range(self.num_stages)]
+
+        self.stages = nn.ModuleList(
+            [
+                FocalNetStage(
+                    config=config,
+                    dim=embed_dim[i_layer],
+                    out_dim=embed_dim[i_layer + 1] if (i_layer < self.num_stages - 1) else None,
+                    input_resolution=(grid_size[0] // (2**i_layer), grid_size[1] // (2**i_layer)),
+                    depth=config.depths[i_layer],
+                    drop_path=dpr[sum(config.depths[:i_layer]) : sum(config.depths[: i_layer + 1])],
+                    focal_level=config.focal_levels[i_layer],
+                    focal_window=config.focal_windows[i_layer],
+                    downsample=FocalNetPatchEmbeddings if (i_layer < self.num_stages - 1) else None,
+                )
+                for i_layer in range(self.num_stages)
+            ]
+        )
+
+        self.gradient_checkpointing = False
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        input_dimensions: Tuple[int, int],
+        output_hidden_states: Optional[bool] = False,
+        output_hidden_states_before_downsampling: Optional[bool] = False,
```

Comment:
What is the purpose of this argument? It seems to only been an option in the encoder but not an option from `FocalNetModel`

Reply:
This model is very similar to SwinModel (it is based on Swin Transformer).

We need the output_hidden_states_before_downsampling argument when the model is used as backbone.

See e.g. [here](https://github.com/huggingface/transformers/blob/73fdc8c5b4509580a047696a4c672a80a8f1710c/src/transformers/models/swin/modeling_swin.py#L1327) where this is used for SwinBackbone.